### ISSUE TITLE
LoRA: performance study

### ISSUE THREAD

**Issue by Andrei-Aksionov** [Lightning-AI/lit-gpt#461]
Hi there 👋 

As a continuation of #448 I decided to create another issue where I'll put results for other small experiments of performance improvements for the LoRA implementation:

- [x] 1. Replace `F.linear` with `F.conv1d`
- [x] 2. Use custom split of tensors inside `F.conv1d` instead of `groups` parameter
- [x] 3. Make `zero_pad` go brr.
- [x] 4. Get rid of transpose calls inside `LoRALinear:forward` call

1. A while back I did a quick [experiment](https://colab.research.google.com/drive/1hOcUyql6mICIjVZwABs1-fxxTT8Iag2s?usp=sharing) comparing the speed of `nn.Linear` vs  `nn.Conv1d` and the convolution variant did win on a GPU.
2. As I stated in #377 manual splitting instead of `groups` parameter might lead to a speed improvement, at least for `Pythia-70m` and `Pythia-410m`. I saw somewhat similar behavior in #448: faster with smaller models and slower with bigger ones. So that needs to be checked.
3. `zero_pad` is slow, perhaps the reason is that for each call we copy files from CPU to GPU
4. In the forward pass of `LoRALinear` class we need to do transpose of `lora_A` as well as `lora_B`:
https://github.com/Lightning-AI/lit-gpt/blob/d5d371417ecb3d3b6c4f30837d8bb7cf2b5310ae/lit_gpt/lora.py#L149

What if to store weights in a transposed form? How much we can gain? I think not too much as time complexity for transposing is    $m\*n$ (where $m$ and $n$ are the number of rows and columns) and since the sizes of LoRA weights are pretty small the effect should be negligible. Plus, it will bring an issue with loading old checkpoints. Nevertheless, want to check it.

**Comment by rasbt** [comment]
@Andrei-Aksionov Thanks for trying! Btw. that's actually a really interesting topic for a blog post

**Comment by Andrei-Aksionov** [comment]
@rasbt I would love to read something about it: tensor manipulations like reshape, transpose, view vs reshape, time complexity of matrix multiplication and how order of it might affect [performance](https://github.com/Lightning-AI/lit-gpt/issues/448), ...
So basically an article explaining that this is not for free and one needs to be cautious otherwise the training might be surprisingly slow.
Might be also relevant to the NeurIPS challenge as people might experiment quite a lot. And for any kagglers out there too.

**Comment by Andrei-Aksionov** [comment]
> 2. Use custom split of tensors inside F.conv1d instead of groups parameter

What's going on here: in #357 you can find an in-depth technical explanation but in short in conv call there is a `groups` argument which lets us to equally split input and weight matrices (e.g. with groups=2 the split will be into two equal parts) and this is equivalent to having multiple conv layers side by side. The problem is that in LoRA with grouped queries the combined QKV matrix consists of unequally sized parts (Q!=K!=V) and thus we need to use a custom way of splitting this matrix.

So what the deal with this experiment? In the code we use `groups` argument when we don't use grouped queries and the manual splitting - when we do. And it's also possible to force using manual approach even if we don't have to. That what I did during PR preparation and noticed that actually the manual approach is faster:
> Tested on a single T4 with Pythia-410m with default parameters (except max_iters which I set to a 1,000). 
I used this model as it's relatively small and doesn't require these changes (as it doesn't use grouped queries) so it allows to compare two approaches: conv1d with groups parameter and manual one with splitting and concatenating tensors:
    1. conv1d with groups finished in 169 seconds
    2. manual approach finished in 128 seconds
Loss after 1,000 iterations is the same in both cases: ~1.801

The difference is massive: ~25% time reduction.

**So the question is: shouldn't then we use manual approach all the time?**

## Act 1. Frustration

I decided to test not only relatively small models but also larger ones so I started with `Nvidia A10G` (remember it), `bf16-true`, LoRA finetune, `1k` iters.
Models are: Pythia 70m-2.8B, StableLM3B and OpenLLaMA-3B.

| $Model$      | $Loss_{curr}$ | $Loss_{forced}$ | $Time_{curr}$ | $Time_{forced}$ | $Micro batch size$ |
|--------------|---------------|-----------------|---------------|-----------------|--------------------|
| Pythia-70m   | 3.848         | 3.8761          | 28.63         | 29.43           | 4                  |
| Pythia-410m  | 1.8500        | 1.8544          | 83.84         | 89.11           | 4                  |
| Pythia-1B    | 1.7301        | 1.7289          | 98.46         | 98.02           | 4                  |
| Pythia-1.4B  | 1.5756        | 1.5708          | 135.88        | 134.88          | 4                  |
| Pythia-2.8B  | 1.3207        | 1.3258          | 227.13        | 223.37          | 4                  |
| OpenLLaMA-3B | 2.7962        | 2.7960          | 163.93        | 162.50          | 2                  |
| StableLM-3B  | 1.5852        | 1.6048          | 240.44        | 238.01          | 4                  |

$_{current}$ - the code from the main branch

$_{forced}$ - force to use the manual split

Can you spot the difference? I mean, yeah, in some cases the forced approach wins, in some - the current one, but for the most part, especially considering that this is just 1k iterations, the results are almost the same.

So what's wrong? Maybe that's because I used `true` precision instead of `mixed` like I did in the experiment for the PR? Let's try it, plus also let's check all three popular precisions: `16-mixed`, `bf16-true`, `bf16-mixed`.

## Act 2. Frustration continues

Here I experimented only with two models (Pythia-70m and Pythia-410m) and with 3 precisions on `A10G` and `1k` iterations.

| $Model$     | $Loss_{curr}$ | $Loss_{forced}$ | $Time_{curr}$ | $Time_{forced}$ | $Precision$ |
|-------------|---------------|-----------------|---------------|-----------------|-------------|
| Pythia-70m  | 2.5772        | 2.5654          | 33.31         | 32.95           | 16-mixed    |
| Pythia-70m  | 3.8757        | 3.8424          | 35.05         | 30.06           | bf16-true   |
| Pythia-70m  | 3.7765        | 3.8714          | 29.99         | 31.22           | bf16-mixed  |
| Pythia-410m | 1.7835        | 1.7819          | 93.93         | 101.27          | 16-mixed    |
| Pythia-410m | 1.8521        | 1.8618          | 91.21         | 90.15           | bf16-true   |
| Pythia-410m | 1.8462        | 1.8420          | 92.04         | 98.44           | bf16-mixed  |

Confused? Me too.
First observation: if to compare bf16-true results from this table with the previous one - we can see fluctuations.
Second observation: still no major difference.

So how in the world I got so big difference with Pythia-410m: 169 vs 129 seconds? 
Ah, here comes the most interesting part!

## Act 3. The old guard strikes back!

And then I recalled that for the PR I did tests on `Nvidia T4`. But it couldn't be the reason, right?

| $Model$     | $Loss_{curr}$ | $Loss_{forced}$ | $Time_{curr}$ | $Time_{forced}$ | $Precision$ | $Microbatchsize$ |
|-------------|---------------|-----------------|---------------|-----------------|-------------|------------------|
| Pythia-70m  | 2.5720        | 2.5694          | 53.63         | 39.76           | 16-mixed    | 4                |
| Pythia-70m  | 2.5122        | 2.5218          | 61.10         | 59.99           | 32-true     | 4                |
| Pythia-410m | 1.7974        | 1.7961          | 171.31        | 127.18          | 16-mixed    | 4                |
| Pythia-410m | 1.7885        | 1.7884          | 348.64        | 357.28          | 32-true     | 4                |
| Pythia-1b   | 3.0006        | 3.0005          | 144.95        | 123.90          | 16-mixed    | 2                |

<p align="center">
<img src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/fe3ce6e8-0d90-49a5-ad74-f592b2e988b0" alt="drawing" width="400" height="200"/>
</p>

Nice, right? If you want stability in your life - ML/AI isn't your choice.

So it looks like `16-mixed` on a `T4` prefers the manual approach, while other precision and the other GPU - don't.
Don't know what is happening on other cards so I would recommend to use the manual approach only when it's necessary - with grouped queries. In other words let's stick to the code that is currently in the main branch 😉.

-----

So, kids, what have we learned today?
If you do any work that targets performance improvements - test your solution on multiple models/GPUs/precisions.
Given how "quickly" you can do it with LLMs - it should be fun.

<p align="center">
<img src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/970d5dc8-5f29-4b54-88c9-7e2487917aaf" alt="drawing" width="200"/>
</p>

**Comment by Andrei-Aksionov** [comment]
> [!NOTE]
> This is a continuation of the comment above where I discovered that on `T4` with `16-mixed` precision and a `manual split` for the `conv1d` operation (forced approach) the training runs faster.

I'm not a huge expert in the PyTorch profiler, far from being it, but I think I've found the reason for this behavior. 
The slowdown is happening in the `zero_pad` method (which we will discuss a bit later).
 
Just take a look at the trace on `T4`, with `16-mixed` precision and the standard grouped `conv1d`: 
<img width="835" alt="Screenshot 2023-09-15 at 3 34 39 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/42732bd0-3dbb-47b0-a074-ef13d8f67a03">

And with same card and parameters, but with manual split:
<img width="835" alt="Screenshot 2023-09-15 at 3 37 37 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/8bcb4e2d-1ac8-4825-9b13-eec42607278b">

With the current (default) conv1d `zero_pad` takes so much time that it dwarfs the trace for `conv1d`. As a reminder: I don't touch `zero_pad` at all during all these experiments.
As a result, with the current approach, `zero_pad` takes `4.18 ms` versus `608 micro seconds` for the manual split.
I think this is because of `cudaMemcpyAsync`:
<img width="835" alt="Screenshot 2023-09-15 at 3 40 18 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/e4615e8b-5028-4b77-94c4-77d1085f8b84">
though I don't know what it `<built-in method tensor of type object at 0x7f7cf40b3ec0>`

And that `cudaMemcpyAsync` is used on `A10G` no matter what precision and what conv1d is used; on T4 only with 16-mixed precision and the manual conv1d split 🤷.
So on T4 `zero_pad` takes much more time than conv1d except for 16-mixed precision and manual conv1d;
on A10G (no matter what precision and conv1d), T4 with 16-mixed precision and manual split `zero_pad` takes slightly more time than `conv1d`.

----

Here are the trace files for anyone who is interested. Maybe one can find something that I couldn't.

[trace_current_16_mixed_t4.json.gz](https://github.com/Lightning-AI/lit-gpt/files/12619489/trace_current_16_mixed_t4.json.gz)
[trace_forced_16_mixed_t4.json.gz](https://github.com/Lightning-AI/lit-gpt/files/12619492/trace_forced_16_mixed_t4.json.gz)

[trace_current_32_true_t4.json.gz](https://github.com/Lightning-AI/lit-gpt/files/12619490/trace_current_32_true_t4.json.gz)
[trace_forced_32_true_t4.json.gz](https://github.com/Lightning-AI/lit-gpt/files/12619493/trace_forced_32_true_t4.json.gz)

[trace_current_16_mixed_a10g.json.gz](https://github.com/Lightning-AI/lit-gpt/files/12619487/trace_current_16_mixed_a10g.json.gz)
[trace_forced_16_mixed_a10g.json.gz](https://github.com/Lightning-AI/lit-gpt/files/12619491/trace_forced_16_mixed_a10g.json.gz)

----

One might ask: I'm sorry, but why `zero_pad`, a function that simply copies values from one tensor to another, a bigger one, takes so much time? In fact, it takes more time than `conv1d` that does a lot of computations??!!

Take a look one more time at the trace:
<img width="835" alt="Screenshot 2023-09-15 at 3 48 59 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/aa83ed09-0a51-4168-ab5d-0a5a52dc6cc2">

Do you see that big fat `cudaStreamSynchronize` along with its friends `_to_copy` and `copy_`?
That's because during each forward pass, we copy data from CPU to GPU.
My suspect is this line:
https://github.com/Lightning-AI/lit-gpt/blob/d38fa3a01218ba9f19c27ae89525abe8a95293a3/lit_gpt/lora.py#L294
Here, for each `zero_pad` call, we create a tensor on a GPU out of the data from `self.lora_ind` (that is stored on a CPU).

At least, this is my theory. In the next comment, I'll put my findings related to this.

**Comment by Andrei-Aksionov** [comment]
> [!NOTE]
> This is a continuation of the comment above where I discovered that `zero_pad` method is too slow presumably because we have a lot of CPU to GPU copying.

In the last image in the comment above we saw that something forces to do a lot of copying from CPU to GPU (and calls `cudaStreamSynchronize` along the way), but how to determine which exact line of the code is doing it? 
We can use `torch.profile.record_function` context manager for it!

Let's do exactly that: split the whole code of the `zero_pad` method into separate functions and wrap each one of them in that context manager. The code will look like this:
```python
with torch.profiler.record_function("1_transpose"):
    x = x.transpose(0, 1)
with torch.profiler.record_function("2_new_zeros"):
    result = x.new_zeros((*x.shape[:-1], self.linear.out_features))  # (64, 64, 384)
with torch.profiler.record_function("3_view"):
    result = result.view(-1, self.linear.out_features)  # (4096, 384)
with torch.profiler.record_function("4_indices"):
    indices = torch.tensor(self.lora_ind, device=result.device)
with torch.profiler.record_function("5_source"):
    source = x.reshape(-1, sum(self.qkv_shapes))
with torch.profiler.record_function("6_index_copy"):
    result = result.index_copy(1, indices, source)  # (4096, 256)
with torch.profiler.record_function("7_view"):
    result = result.view((*x.shape[:-1], self.linear.out_features))
with torch.profiler.record_function("8_transpose"):
    result = result.transpose(0, 1)  # (64, 64, 384)
return result
```

> [!NOTE]
> Here I'll use the default conv1 (current approach) on `T4` with `16-mixed` precision, the one where `zero_pad` was humongous.  

Now let's check what PyTorch profiling tells us:
<img width="1040" alt="Screenshot 2023-09-15 at 8 06 45 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/22603c45-863b-4dee-8f1e-bee2049b81f7">
Yep, this is exactly what I suspected: `4_indices` where we create a new tensor on a GPU out of `self.lora_ind` list that is stored on a CPU.

If to store `self.lora_ind` as a tensor on a GPU instead of python list and do it inside `__init__` method, then we can see that it indeed helps:
<img width="1040" alt="Screenshot 2023-09-15 at 8 08 51 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/57b0e42b-3e67-418e-acd9-f454323f097c">

As a result, before `zero_pad` took `4.2 ms` and now - `418 μs`.

But it is still longer than `conv1d` that takes `343 μs`.
What exactly takes so much time? 
<img width="1168" alt="Screenshot 2023-09-15 at 8 18 09 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/6027a3cb-79c7-4478-af02-4b306282c41a">
1. Creation of the result by calling `x.new_zeros`
2. Reshape of the input x
3. Index_copy

And I'm not sure that we can do anything with that. 
But that doesn't mean that I won't try 😄.

**Comment by rasbt** [comment]
Thanks for all the very detailed and thorough work! I wanted to say that I really appreciate this! It's on my list of things I want to read carefully so that I can say something intelligently. I hope I can carve out some more time in the upcoming week!

**Comment by Andrei-Aksionov** [comment]
> 4. Get rid of transpose calls inside LoRALinear:forward call

Based on the trace (T4, Pythia-410m) transposing takes `~9-12 μs` out of `~530 μs` for the whole forward pass of the `LoRALinear` class. In percents it's `1.73-2.30%`.

Given that this change will cause incompatibility with old checkpoints, I think the change simply doesn't worth it.
<img width="881" alt="Screenshot 2023-10-09 at 6 12 26 PM" src="https://github.com/Lightning-AI/lit-gpt/assets/58434077/e9c1fb6a-53d8-4065-8fcb-dfa252f96b9a">

**Comment by NamburiSrinath** [comment]
Hi @Andrei-Aksionov , @rasbt 

Can I know the steps to profile the code in Litgpt repo? You have produced some figures above, can you help us understand how you were able to do that.

Currently, I can run the code without any issue (eg: `litgpt pretrain pythia-14m --config xxx.yaml --out_dir out/xxx`) 

Basically, I wanted to know whether attention blocks or FFNs or embeddings are taking most of the memory while I am pretraining a model (there're some code changes at my end and I wrap the litgpt as a module, but the above command works). 

The README doesn't give more details on this and I am interested in this setup.
